File: p4/p4.py
# Problem 4
#
# A palindromic number reads the same both ways. The largest palindrome 
# made from the product of two 2-digit numbers is 9009 = 91  99.
#
# Find the largest palindrome made from the product of two 3-digit numbers.

max_product = 0

for i in range(100, 1000):
   for j in range(100, 1000):
       product = i * j
       product_str = str(product)
       if (product_str == product_str[::-1]): 
           max_product = max(max_product, product)

print max_product
      

File: csharp/P1_to_10/P4.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ProjectEuler.P1_to_10
{
    /// <summary>
    /// https://projecteuler.net/problem=4
    /// </summary>
    internal class P4
    {
        /// <summary>
        /// Finds the largest product of two 3-digit numbers that is also a palindrome number.
        /// </summary>
        public int Solve()
        {
            var products = from x in Enumerable.Range(100, 899)
                           from y in Enumerable.Range(100, 899)
                           select x * y;

            return products.Where(IsPalindromeNumber).Max();
        }

        /// <summary>
        /// Returns true if the number is a palindrome number, otherwise false.
        /// </summary>
        private bool IsPalindromeNumber(int number)
        {
            var str = number.ToString();
            for (var i = 0; i < str.Length/2; i++)
            {
                if (str[i] != str[str.Length - i - 1])
                {
                    return false;
                }
            }

            return true;
        }
    }
}
